homeassistant/components/wyoming/conversation.py

```diff
@@ -0,0 +1,207 @@
+"""Support for Wyoming intent recognition services."""
+
+import logging
+
+from wyoming.asr import Transcript
+from wyoming.client import AsyncTcpClient
+from wyoming.handle import Handled, NotHandled
+from wyoming.info import HandleProgram, IntentProgram
+from wyoming.intent import Intent, NotRecognized
+
+from homeassistant.components import conversation
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import intent
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util import ulid
+
+from .const import DOMAIN
+from .data import WyomingService
+from .error import WyomingError
+from .models import DomainDataItem
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Wyoming conversation."""
+    item: DomainDataItem = hass.data[DOMAIN][config_entry.entry_id]
+    async_add_entities(
+        [
+            WyomingConversationEntity(config_entry, item.service),
+        ]
+    )
+
+
+class WyomingConversationEntity(
+    conversation.ConversationEntity, conversation.AbstractConversationAgent
+):
+    """Wyoming conversation agent."""
+
+    _attr_has_entity_name = True
+
+    def __init__(
+        self,
+        config_entry: ConfigEntry,
+        service: WyomingService,
+    ) -> None:
+        """Set up provider."""
+        super().__init__()
+
+        self.entry = config_entry
+        self.service = service
+
+        self._intent_service: IntentProgram | None = None
+        self._handle_service: HandleProgram | None = None
+
+        for maybe_intent in self.service.info.intent:
+            if maybe_intent.installed:
+                self._intent_service = maybe_intent
+                break
+
+        for maybe_handle in self.service.info.handle:
+            if maybe_handle.installed:
+                self._handle_service = maybe_handle
+                break
+
+        model_languages: set[str] = set()
+
+        if self._intent_service is not None:
+            for intent_model in self._intent_service.models:
+                if intent_model.installed:
+                    model_languages.update(intent_model.languages)
+
+            self._attr_name = self._intent_service.name
+            self._attr_supported_features = (
+                conversation.ConversationEntityFeature.CONTROL
+            )
+        elif self._handle_service is not None:
+            for handle_model in self._handle_service.models:
+                if handle_model.installed:
+                    model_languages.update(handle_model.languages)
+
+            self._attr_name = self._handle_service.name
+        else:
+            raise ValueError("No intent or handle service")
```

Comment:
Adding an entity cannot just raise. It should not add the entity unless there is something that can handle it .